### PR TITLE
INS-48489 Bigtable filter plugin

### PR DESCRIPTION
**Pull request by pankajkumaribm**

### CODE REVIEW

**Review by taees-eimouri — APPROVED** [review]

**Review by chirag-ibm — APPROVED** [review]

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
+# BigTable-Guardium Logstash filter plug-in
+### Meet BigTable
+* Tested versions: V2
+* Environment: Google Cloud Platform (GCP)
+* Supported inputs: Pub/Sub (pull)
+* Supported Guardium versions:
+    * Guardium Data Protection: 12.1 and above
+    * Guardium Insights: 3.3 and above
+
+This is a [Logstash](https://github.com/elastic/logstash) filter plug-in for the universal connector that is featured in IBM Security Guardium. It parses a GCP (Google Cloud Platform) event logs into a [Guardium record](https://github.com/IBM/universal-connectors/blob/main/common/src/main/java/com/ibm/guardium/universalconnector/commons/structures/Record.java) instance (which is a standard structure made out of several parts). The information is then sent over to Guardium. Guardium records include the accessor (the person who tried to access the data), the session, data, and exceptions. If there are no errors, the data contains details about the query "construct". The construct details the main action (verb) and collections (objects) involved. As of now, the BigTable plug-in only supports Guardium Data Protection.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:10]: This statement is incorrect - As of now, the BigTable plug-in only supports Guardium Data Protection.
Change to - "The BigTable-Guardium Logstash filter plug-in supports Guardium Data Protection and Guardium Insights."

Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:10]: Change "It parses a GCP (Google Cloud Platform) event logs into a Guardium record" to 
""It parses GCP (Google Cloud Platform) event logs into a Guardium record"

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+### Note
+This plug-in contains a runtime dependency of Logstash Google PubSub input plug-in (version ~> 1.2.1, i.e. at least 1.2.1).
+
+This version is compliant with Guardium Data Protection v12.1 and above. Please refer to the [input plug-in repository](https://github.com/IBM/universal-connectors/tree/main/input-plugin/logstash-input-google-pubsub) for more information.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:15]: Change "Please refer to the [input plug-in repository](https://github.com/IBM/universal-connectors/tree/main/input-plugin/logstash-input-google-pubsub) for more information." to
For more information, refer [input plug-in repository](https://github.com/IBM/universal-connectors/tree/main/input-plugin/logstash-input-google-pubsub).

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+## Configuring BigTable on GCP
+### BigTable Setup
+1. [Prerequisites](https://cloud.google.com/bigtable/docs/quickstarts/quickstart-cloud-console)
+2. BigTable is automatically enabled in new projects. To activate BigTable in an existing project, enable the BigTable API.  Please refer to [Enable the BigTable API](https://console.cloud.google.com/marketplace/product/google/bigtable.googleapis.com) for more information.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:20]: Change "To activate BigTable in an existing project, enable the BigTable API.  Please refer to [Enable the BigTable API](https://console.cloud.google.com/marketplace/product/google/bigtable.googleapis.com) for more information." to
 "To activate BigTable in an existing project, refer Enable the BigTable API](https://console.cloud.google.com/marketplace/product/google/bigtable.googleapis.com).

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+3. Create a BigTable instance that stores the data. Please refer to [Create a Dataset](https://cloud.google.com/bigtable/docs/creating-instance) for more information.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:21]: For more information, refer [Create a Dataset](https://cloud.google.com/bigtable/docs/creating-instance).

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+4. Create a table. Please refer to [Create Table](https://cloud.google.com/bigtable/docs/samples/bigtable-hw-create-table) for more information.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:22]: For more information, refer  [Create Table](https://cloud.google.com/bigtable/docs/samples/bigtable-hw-create-table).

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+5. Query table data. Please refer to [Query Table Data](https://cloud.google.com/bigquery/docs/external-data-bigtable) for more information.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:23]: For more information, refer [Query Table Data](https://cloud.google.com/bigquery/docs/external-data-bigtable)

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+### Permissions and Roles Details
+
+##### Grant Permission Required for log view/ download
+
+User can view and download the generated logs.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:29]: You can view and download the generated logs.
You need the following identity and access management roles to view and download the generated logs:
* View logs:
roles/logging.viewer (Logs Viewer) 
roles/logging.privateLogViewer (Private Logs Viewer)
* Download logs:
roles/logging.admin (Logging Admin)
roles/logging.viewAccessor (Logs View Accessor)

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+The following Identity and Access Management roles are required to view and download logs:
+
+* To view logs:
+roles/logging.viewer (Logs Viewer)
+roles/logging.privateLogViewer (Private Logs Viewer)
+* To download logs:
+roles/logging.admin (Logging Admin)
+roles/logging.viewAccessor (Logs View Accessor)
+
+### Create a topic in Pub/Sub
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:39]: Can you change the topic title for task topics to use gerunds. For example: Creating a topic in Pub/Sub, Creating a subscription in Pub/Sub. Please make these changes for other topics In **Configuring BigTable on GCP**

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+### Create a log sink in Pub/Sub
+* In the Cloud Console, go to the ```Logging > Log Router``` page.
+* Click ```Create sink```.
+* In the Sink details panel, enter the following details:
+* ```Sink name```: Provide an identifier for the sink. Note that after you create the sink you cannot rename it. However, you can delete a sink and create a new one.
+* ```Sink description (optional)```: Describe the purpose or use case for the sink.
+* In the ```Sink destination``` panel, select the Cloud ```Pub/Sub topic``` as sink service and select the topic created in previous steps.
+* Choose logs to include in the sink in the Build inclusion filter panel.
+* You can filter the logs by log name, resource, and  severity.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:59]: Remove the bullet points for  You can filter the logs by log name, resource, and severity. Multi-region

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+Multi-region
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:60]: Can you filter logs by multi-region.
Then, change to - You can filter the logs by log name, resource, severity, and multi-region.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+* In cases of multiple regions, you need to do the same set of configurations per each region.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:61]: Remove the bullet point for this sentence.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+Based on the region, different configuration files will be used for the input plug-in
+
+#### Set destination (TOPIC & SUBSCRIPTION) permissions
+
+To set permissions for the log sink to route to its destination, do the following:
+* Obtain the sink's writer identity—an email address—from the new sink.
+   1. Go to the ```Log Router``` page, and select ```menu```  > ```View sink details```.
+   2. The writer identity appears in the Sink details panel.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:69]: This is not a step. Remove the numbered list.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+* If you have owner access to the destination:
+   1. Add the sink's writer identity to topic >>>>
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:71]: Why is period missing in all the steps?

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+      - Navigate to the Topic created in the earlier steps
+      - Click on SHOW INFO panel
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:73]: Can we remove "on" and change to:
Click SHOW INFO panel.
Click ADD PRINCIPAL.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+
+   2. Add the sink's writer identity to subscription >>>>
+      - Navigate to the Subscription
+      - Click on SHOW INFO panel
+      - Click on ADD PRINCIPAL
+      - Paste writer identity in the New Principals
+      - Give it the subscriber role
+
+### Create service account credentials
+* Go to the Service accounts section of the IAM & Admin console.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:86]: Change the bullet list to numbered list as they are to followed in a specific order.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+
+## Viewing the Audit logs
+
+The inclusion filter mentioned above will be used to view the Audit logs in the GCP Logs Explorer.
+
+### Supported audit logs
+  1. BigTableAudit - `ACTIVITY`, `DATA_ACCESS` logs
+  2. BigTable Log - `CREATEINSTANCE`, `DELETEINSTANCE`, `UPDATEINSTANCE`, `CREATECLUSTER`, `DELETECLUSTER`, `UPDATECLUSTER`, `CREATETABLE`, `DELETETABLE`, `MODIFYCOLUMNFAMILIES`, `EXECUTEQUERY`, `LISTCLUSTERS`, `LISTINSTANCES`. 
+
+## 4. Limitations
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:111]: Why is "Limitations" numbered "4" when none of the other topics are numbered

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+```    
+protoPayload.serviceName="bigtableadmin.googleapis.com" OR protoPayload.serviceName="bigtable.googleapis.com"
+```
+
+## Viewing the Audit logs
+
+The inclusion filter mentioned above will be used to view the Audit logs in the GCP Logs Explorer.
+
+### Supported audit logs
+  1. BigTableAudit - `ACTIVITY`, `DATA_ACCESS` logs
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:108]: Do we need a numbered list. Can we change to bullet list.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+  2. BigTable Log - `CREATEINSTANCE`, `DELETEINSTANCE`, `UPDATEINSTANCE`, `CREATECLUSTER`, `DELETECLUSTER`, `UPDATECLUSTER`, `CREATETABLE`, `DELETETABLE`, `MODIFYCOLUMNFAMILIES`, `EXECUTEQUERY`, `LISTCLUSTERS`, `LISTINSTANCES`. 
+
+## 4. Limitations
+1. If no information regarding certain fields is available in the logs, those fields will not be mapped.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:112]: Change "If no information regarding certain fields is available in the logs, those fields will not be mapped" to
The fields whose information is not available in the logs are not mapped.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+
+## 4. Limitations
+1. If no information regarding certain fields is available in the logs, those fields will not be mapped. 
+2. Exception object will be prepared based on severity of the logs.
+3. The data model size is limited to 10 GB per table. If you have a 100 GB reservation per project per location, BigTable BI Engine limits the reservation per table to 10 GB. The rest of the available reservation is used for other tables in the project.
+4. BigTable cannot read the data in parallel if you use gzip compression. Loading compressed JSON data into BigTable is slower than loading uncompressed data.
+5. You cannot include both compressed and uncompressed files in the same load job.
+6. JSON data must be newline delimited. Each JSON object must be on a separate line in the file.
+7. The maximum size for a gzip file is 4 GB.
+8. Log messages have a size limit of 100K bytes
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:119]: Period missing at the end of the sentence.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+9. The Audit/Data access log doesn't contain a server IP. The default value is set 0.0.0.0 and can be in IPV4 or IPV6 format.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:120]: Change to "The default value is set to 0.0.0.0 and can be in IPV4 or IPV6 format.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+10. The following important fields cannot be mapped, as there is no information regarding these fields in the logs:
+    - Source program 
+    - OS User
+    - Client HostName 
+11.  `serverHostName` pattern for BigTable GCP :
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:125]: This does not look like a limitation. Is this sentence incomplete?

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+`project-id_bigtableadmin.googleapis.com` | `project-id_bigtable.googleapis.com`.
+12. While using GCP, excluding the web UI, duplicate entries may appear in both the reports and audit logs.
+13. Bigtable uses two different service names (`bigtable.googleapis.com` & `bigtableadmin.googleapis.com`) depending on the tasks being performed. This will result in two distinct S-TAP Host entries.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:128]: Change to - This results in two distinct S-TAP host entries.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+Edit the Sink via *Logs Router > Sink Inclusion Filter*:
+#### Description
+The purpose of this inclusion filter is to exclude unnecessary logs and include required logs with resource types and metadata only from BigTable.
+```    
+protoPayload.serviceName="bigtableadmin.googleapis.com" OR protoPayload.serviceName="bigtable.googleapis.com"
+```
+
+## Viewing the Audit logs
+
+The inclusion filter mentioned above will be used to view the Audit logs in the GCP Logs Explorer.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:105]: Change to - The inclusion filter mentioned above is used to view the Audit logs in the GCP Logs Explorer.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+
+### Supported audit logs
+  1. BigTableAudit - `ACTIVITY`, `DATA_ACCESS` logs
+  2. BigTable Log - `CREATEINSTANCE`, `DELETEINSTANCE`, `UPDATEINSTANCE`, `CREATECLUSTER`, `DELETECLUSTER`, `UPDATECLUSTER`, `CREATETABLE`, `DELETETABLE`, `MODIFYCOLUMNFAMILIES`, `EXECUTEQUERY`, `LISTCLUSTERS`, `LISTINSTANCES`. 
+
+## 4. Limitations
+1. If no information regarding certain fields is available in the logs, those fields will not be mapped. 
+2. Exception object will be prepared based on severity of the logs.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:113]: Change to - Exception object is prepared based on severity of the logs.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+10. The following important fields cannot be mapped, as there is no information regarding these fields in the logs:
+    - Source program 
+    - OS User
+    - Client HostName 
+11.  `serverHostName` pattern for BigTable GCP :
+`project-id_bigtableadmin.googleapis.com` | `project-id_bigtable.googleapis.com`.
+12. While using GCP, excluding the web UI, duplicate entries may appear in both the reports and audit logs.
+13. Bigtable uses two different service names (`bigtable.googleapis.com` & `bigtableadmin.googleapis.com`) depending on the tasks being performed. This will result in two distinct S-TAP Host entries.
+14. Multiple sessions from the same Bigtable instance may result in multiple S-TAP entries. 
+15. The BigTable audit log doesn’t include login failed logs, so these will not appear in the guardium LOGIN_FAILED report.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:130]: Change to - The BigTable audit log doesn’t include login failed logs. So, these logs do not appear in the Guardium LOGIN_FAILED report.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+## Configuring the BigTable filter in Guardium
+The Guardium universal connector is the Guardium entry point for native audit/data_access logs. The Guardium universal connector identifies and parses the received events, and converts them to a standard Guardium format. The output of the Guardium universal connector is forwarded to the Guardium sniffer on the collector, for policy and auditing enforcements. Configure Guardium to read the native audit/data_access logs by customizing the BigTable template.
+
+### Before you begin
+* Configure the policies you require. See [policies](/docs/#policies) for more information.
+* You must have permission for the S-Tap Management role. The admin user includes this role by default
+* Download the [logstash-filter-big_table_guardium_filter](gdp-pubsub-bigtable-package/logstash-filter-big_table_guardium_filter.zip) plug-in.
+
+### Procedure
+1. On the collector, go to Setup > Tools and Views > Configure Universal Connector.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:141]: Change to - On the collector, go to ```Setup``` > ```Tools and Views``` > ```Configure Universal Connector```.

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+2. Enable the universal connector if it is disabled.
+3. Click Upload File and select the offline  [logstash-filter-big_table_guardium_filter](gdp-pubsub-bigtable-package/logstash-filter-big_table_guardium_filter.zip) plug-in. After it is uploaded, click ```OK```.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:143]: Change to - Click  ```Upload File```

File: filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md [hunk 28]
```diff
@@ -0,0 +1,152 @@
...
+4. Click ```Upload File``` and select the key.json file. After it is uploaded, click ```OK```.
+5. Click the Plus sign to open the Connector Configuration dialog box.
+6. Type a name in the Connector name field.
+7. Update the input section to add the details from the [pubsub_big_table.conf](gdp-pubsub-bigtable-package/pubsub_big_table.conf) file's input part, omitting the keyword "input{" at the beginning and its corresponding "}" at the end.
+8. Update the filter section to add the details from the [pubsub_big_table.conf](gdp-pubsub-bigtable-package/pubsub_big_table.conf) file's filter part, omitting the keyword "filter{" at the beginning and its corresponding "}" at the end.
+9. The 'type' fields should match in the input and filter configuration sections. This field should be unique for every individual connector added.
+10. Click ```Save```. Guardium validates the new connector and displays it in the Configure Universal Connector page.
+11. After the offline plug-in is installed and the configuration is uploaded and saved in the Guardium machine, restart the Universal Connector using the Disable/Enable button.
```
Comment (VanajaGP) [inline filter-plugin/logstash-filter-pubsub-bigtable-guardium/README.md:151]: Change to ```Disable/Enable``` button.

**Review by VanajaGP — COMMENTED** [review]
Reviewed the readme. Reach out to me for any clarifications.

**Comment by VanajaGP** [comment]
All changes look good.

**Review by taees-eimouri — APPROVED** [review]

**Review by hadarkorny — CHANGES_REQUESTED** [review]
Please check these tasks before merging this PR. 
- [x] Tested internal build script for offline packages. 

- [x] Documentation reviewed by @ShalakaKulkarni15.

**Review by doryo — APPROVED** [review]

**Comment by ShalakaKulkarni15** [comment]
@hadarkorny, @VanajaGP had already reviewed this PR in my absence.

**Review by VanajaGP — APPROVED** [review]
Changes look good

**Pull request merged**